src/zope/locking/tests.py

```diff
@@ -1,54 +1,54 @@
+#############################################################################
+#
+# Copyright (c) 2018 Zope Foundation and Contributors.
+# All Rights Reserved.
+#
+# This software is subject to the provisions of the Zope Public License,
+# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
+# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
+# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
+# FOR A PARTICULAR PURPOSE.
+#
+##############################################################################
+
+import doctest
 import unittest
-
-import persistent.interfaces
-import ZODB.DB
-import ZODB.MappingStorage
-import transaction
-import zope.app.keyreference.interfaces
-import zope.app.keyreference.persistent
-import zope.app.testing.placelesssetup
-import zope.component
-import zope.event
-
 import zope.locking.testing
 
-from zope.testing import doctest
 
+def test_suite():
+
+    layer = zope.locking.testing.layer
 
-def setUp(test):
-    zope.app.testing.placelesssetup.setUp(test)
-    db = test.globs['db'] = ZODB.DB(ZODB.MappingStorage.MappingStorage())
-    test.globs['conn'] = db.open()
-    test.globs['Demo'] = zope.locking.testing.Demo
-    zope.component.provideAdapter(zope.locking.testing.DemoKeyReference)
-    zope.component.provideAdapter(
-        zope.app.keyreference.persistent.KeyReferenceToPersistent,
-        [persistent.interfaces.IPersistent],
-        zope.app.keyreference.interfaces.IKeyReference)
-    events = test.globs['events'] = []
-    zope.event.subscribers.append(events.append)
+    def get_connection():
+        return layer.db.open()
 
-def tearDown(test):
-    zope.app.testing.placelesssetup.tearDown(test)
-    transaction.abort()
-    test.globs['conn'].close()
-    test.globs['db'].close()
-    events = test.globs.pop('events')
-    assert zope.event.subscribers.pop().__self__ is events
-    del events[:] # being paranoid
+    def get_db():
+        return layer.db
 
-def test_suite():
-    return unittest.TestSuite((
+    suite = unittest.TestSuite((
         doctest.DocFileSuite(
             'README.txt',
-            setUp=setUp, tearDown=tearDown),
+            optionflags=doctest.IGNORE_EXCEPTION_DETAIL,
+            globs=dict(
+                get_connection=get_connection,
+                get_db=get_db
+            )),
```

Comment:
I'm assuming the test layer takes care of tearDown now?

Reply:
Yes.